text/0000-recover-with-receiver.md

```diff
@@ -0,0 +1,153 @@
+as well as automatic receiver recovery. In some cases, these are
+equivalent. If we have a single variable `x: T iso`, then we can
+temporarily use it as another capability inside a recover block,
+with something like:
+```
+x = recover
+   let x_ref = consume ref x
+   // do something with x ...
+   x_ref
+end
+```
+Alternatively, if the action being taken is precisely a ref method call,
+then automatic receiver recovery can be used if the
+arguments and return types meet the isolation guarantees for `x`.
+```
+x.foo(y, z)
+```
+But this can't be used for every type of action, it needs to be those
+actions thought of by the original class developer. We can add these methods,
+but this is anti-modular.
+
+This automatic receiver recovery syntax also works for expressions more complicated than a single variable of course.
+The recover block is less flexible. The recover block method can be used only when the thing being modified is a mutable location.
+It can be used with var fields, but not with let or embed. It can be used when we have update methods, but not with getters alone.
+```
+// defined elsehwere
+class Foo
+   fun box getSomething(): this->Bar ref
+   ...
+   fun box values(): this->FooIterator ref
+
+class FooHolder
+   embed foo: Foo iso = Foo
+   // or let
+
+   fun ref doSomethingWithFoo() =>
+      // error, iso->ref = tag
+      foo.getSomething().somethingElse()
+
+      // try to recover to use foo as ref: error, can't assign
+      foo = recover
+         ... consume foo
+      end
+   end
+```
+We might also have read-only methods. Imagine we take in an Iter over iso objects. We don't want to be coupled to
+the class used, such as Array, and allow a generic, potentially chained iterator.
+```
+class UsesIter[T: SomeInterface]
+   fun process(iter: Iterator[T]) =>
+      // want to call a few complicated methods on T
+      // if T might be unique, we can't store in a variable,
+      // so we want to recover, but we can't!
+
+      // error, not a subtype
+      let next: T = iter.next()?
+
+      // ????
+      iter.next() = recover
+         ...
+      end
+
+      // works... but only if we can
+      // express *all* of the things we want
+      // to do as multiple methods
+      // still anti-modular!
+      iter.next()?.foo().>bar()
+   end
+```
+
+This RFC will add a syntax to expand the design of recover blocks to allow a receiver, subsuming automatic receiver recovery.
+In both cases above, the recover with receiver may be used in order to temporarily use these values as ref, allowing free
+usage of methods, without requiring that the methods were defined ahead of time in the interface or class, and without
+requiring extra potentially erroring accesses or allocating and swapping new values via update methods.
+
+# Detailed design
+
+We will add new syntactic forms to allow recover blocks based around an existing receiver expression.
+
+```
+e.recover as x =>
+   e
+end
```

Comment:
There have been discussions in the past about syntax in which we have decided to try to avoid "overloading" a keyword or symbol with multiple meanings. In this case I think we should avoid using `as` for this syntax, since it is entirely different from how we use `as` for runtime type matching with the right side being a type name.

Not that I'm saying I like everything about Pony syntax as it is today (I don't), but I'm just raising this comment for consistency purposes.

In Pony syntax I think the most applicable symbol/keyword to use here would be `|` (as used in `match` blocks).

```suggestion
e.recover | x =>
   x
end
```

Also I think you meant to use `x` inside the block instead of `e`, so I included that change in this suggestion as well.

Reply:
I applied this manually, but the use of `e` denotes an expression, hence why `x` is not used. These are grammatical definitions, not examples.

Reply:
@jasoncarr0 that confused me as well; I think `e.recover | x => ... f(x) end` (or similar) would be more readable.

Reply:
What about `e[x]`? would that make it clear, or `... x ...`. Of course at the end of the day this is a pretty meaningless subset of the RFC, and I can go to what makes it clear, but on a personal level those ones feel incorrect.